src/h3/mod.rs

```diff
@@ -26,4 +26,830 @@
+        // Some of these are later formatted back into a complete URI
+        let mut method = String::new();
+        let mut scheme = String::new();
+        let mut authority = String::new();
+        let mut path = String::new();
+
+        for hdr in decoder.decode(hdr_block).unwrap() {
+            // trace!("Header field - {}:{}", hdr.0, hdr.1);
+
+            match hdr.name() {
+                ":method" => {
+                    method = hdr.value().to_string();
+                },
+                ":scheme" => {
+                    scheme = hdr.value().to_string();
+                },
+                ":authority" => {
+                    authority = hdr.value().to_string();
+                },
+                ":path" => {
+                    path = hdr.value().to_string();
+                },
+                _ => {
+                    req.headers_mut().insert(
+                        http::header::HeaderName::from_bytes(
+                            hdr.name().as_bytes(),
+                        )
+                        .unwrap(),
+                        http::header::HeaderValue::from_str(hdr.value()).unwrap(),
+                    );
+                },
+            }
+        }
+
+        let uri = format!("{}://{}{}", scheme, authority, path);
+
+        *req.method_mut() = method.parse().unwrap();
+        *req.version_mut() = http::Version::HTTP_2;
+        *req.uri_mut() = uri.parse::<Uri>().unwrap();
+
+        // debug!("Prepared request {:?}", req);
+
+        req
+    }
+
+    pub fn resp_hdrs_from_qpack(
+        decoder: &mut qpack::Decoder, hdr_block: &mut Vec<u8>,
+    ) -> http::Response<()> {
+        let mut resp: Response<()> = Response::default();
+
+        // TODO make pseudo header parsing more efficient.
+        let mut status = String::new();
+
+        for hdr in decoder.decode(hdr_block).unwrap() {
+            // trace!("Header field - {}:{}", hdr.0, hdr.1);
+
+            match hdr.name() {
+                ":status" => {
+                    status = hdr.value().to_string();
+                },
+                _ => {
+                    resp.headers_mut().insert(
+                        http::header::HeaderName::from_bytes(
+                            hdr.name().as_bytes(),
+                        )
+                        .unwrap(),
+                        http::header::HeaderValue::from_str(hdr.value()).unwrap(),
+                    );
+                },
+            }
+        }
+
+        *resp.status_mut() = StatusCode::from_bytes(status.as_bytes()).unwrap();
+        *resp.version_mut() = http::Version::HTTP_2;
+
+        // debug!("Prepared response {:?}", resp);
+
+        resp
+    }
+}
+
+impl H3Connection {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(config: &mut H3Config, is_server: bool) -> Result<Box<H3Connection>> {
+        let root = String::clone(&config.root_dir); // TODO shouldn't need to clone here
+
+        Ok(Box::new(H3Connection {
```

Comment:
We don't need to box the h3 connection (we do this for the transport connection because it's required in that specific case).